src/sql/src/plan/statement/dml.rs

```diff
@@ -216,7 +217,21 @@ fn plan_select_inner(
     } = query::plan_root_query(scx, select.query.clone(), QueryLifetime::OneShot)?;
     expr.bind_parameters(params)?;
 
-    // A top-level limit cannot be data dependent so eagerly evaluate it.
+    // OFFSET clauses in `expr` should become constants with the above binding of parameters.
+    // Let's check this and simplify them to literals.
+    expr.try_visit_mut_pre(&mut |expr| {
+        if let HirRelationExpr::TopK { offset, .. } = expr {
+            let offset_value = offset_into_value(offset.take())?;
+            *offset = HirScalarExpr::literal(Datum::Int64(offset_value), ScalarType::Int64);
+        }
+        Ok::<(), PlanError>(())
+    })?;
```

Comment:
Why do we simply here for offsets but not limits? It seems like we should be simplifying too? I might be misunderstanding though!

Reply:
`LIMIT` clauses in `expr` need not be simplified into a simple value to be able to lower them, see https://github.com/MaterializeInc/materialize/pull/23276
So, we can keep working with them even if they can't be simplified to a value. If they can be simplified to a value, then the optimizer will do that later.

(But even `LIMIT` clauses need to be simplified into a constant in the `RowSetFinishing`, which happens a few lines below.)

I'll add a code comment to note this.